README.md






<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
  <link rel="dns-prefetch" href="https://github.githubassets.com">
  <link rel="dns-prefetch" href="https://avatars0.githubusercontent.com">
  <link rel="dns-prefetch" href="https://avatars1.githubusercontent.com">
  <link rel="dns-prefetch" href="https://avatars2.githubusercontent.com">
  <link rel="dns-prefetch" href="https://avatars3.githubusercontent.com">
  <link rel="dns-prefetch" href="https://github-cloud.s3.amazonaws.com">
  <link rel="dns-prefetch" href="https://user-images.githubusercontent.com/">



  <link crossorigin="anonymous" media="all" integrity="sha512-FG+rXqMOivrAjdEQE7tO4BwM1poGmg70hJFTlNSxjX87grtrZ6UnPR8NkzwUHlQEGviu9XuRYeO8zH9YwvZhdg==" rel="stylesheet" href="https://github.githubassets.com/assets/frameworks-146fab5ea30e8afac08dd11013bb4ee0.css" />
  
    <link crossorigin="anonymous" media="all" integrity="sha512-whtr9xYX7utnpWsNSLW7XLm7eJONfryMwfwxIH2SpIRKCZbx4aryDfn/HGMFI5Fee7dogmqmtqvPPh13+2HW2Q==" rel="stylesheet" href="https://github.githubassets.com/assets/github-c21b6bf71617eeeb67a56b0d48b5bb5c.css" />
    
    
    
    


  <meta name="viewport" content="width=device-width">
  
  <title>awesome-readme/readme.md at master · matiassingers/awesome-readme</title>
    <meta name="description" content="A curated list of awesome READMEs. Contribute to matiassingers/awesome-readme development by creating an account on GitHub.">
    <link rel="search" type="application/opensearchdescription+xml" href="/opensearch.xml" title="GitHub">
  <link rel="fluid-icon" href="https://github.com/fluidicon.png" title="GitHub">
  <meta property="fb:app_id" content="1401488693436528">

    <meta name="twitter:image:src" content="https://avatars2.githubusercontent.com/u/938453?s=400&amp;v=4" /><meta name="twitter:site" content="@github" /><meta name="twitter:card" content="summary" /><meta name="twitter:title" content="matiassingers/awesome-readme" /><meta name="twitter:description" content="A curated list of awesome READMEs. Contribute to matiassingers/awesome-readme development by creating an account on GitHub." />
    <meta property="og:image" content="https://avatars2.githubusercontent.com/u/938453?s=400&amp;v=4" /><meta property="og:site_name" content="GitHub" /><meta property="og:type" content="object" /><meta property="og:title" content="matiassingers/awesome-readme" /><meta property="og:url" content="https://github.com/matiassingers/awesome-readme" /><meta property="og:description" content="A curated list of awesome READMEs. Contribute to matiassingers/awesome-readme development by creating an account on GitHub." />

  <link rel="assets" href="https://github.githubassets.com/">
  <link rel="web-socket" href="wss://live.github.com/_sockets/VjI6NTE4NTEyMTg4OjFlYjZmYmQzYWFiNTVjOTA5MmJiYzU5NTEzZmY4YmMwNTI5MzJhYmM2ZTdiZjNlMjg4MjJkNzk2Zjk0MTM3MDM=--8d16007df2b3ad653e546c05f8d69b714c260daf">
  <link rel="sudo-modal" href="/sessions/sudo_modal">

  <meta name="request-id" content="323D:6F2C:5339A2:86243A:5E896EAA" data-pjax-transient="true" /><meta name="html-safe-nonce" content="66972f4baeeb32de0c6346908783be412c3a6f47" data-pjax-transient="true" /><meta name="visitor-payload" content="eyJyZWZlcnJlciI6Imh0dHBzOi8vZ2l0aHViLmNvbS9tYXRpYXNzaW5nZXJzL2F3ZXNvbWUtcmVhZG1lIiwicmVxdWVzdF9pZCI6IjMyM0Q6NkYyQzo1MzM5QTI6ODYyNDNBOjVFODk2RUFBIiwidmlzaXRvcl9pZCI6IjgzMTg2MDkwNDU0OTg3MTQ3NTciLCJyZWdpb25fZWRnZSI6ImlhZCIsInJlZ2lvbl9yZW5kZXIiOiJpYWQifQ==" data-pjax-transient="true" /><meta name="visitor-hmac" content="1d655361ad70ab1dafaf6d24c5e76e3a9a292a02cf237fe955a033158860f02d" data-pjax-transient="true" />



  <meta name="github-keyboard-shortcuts" content="repository,source-code" data-pjax-transient="true" />

  

  <meta name="selected-link" value="repo_source" data-pjax-transient>

    <meta name="google-site-verification" content="KT5gs8h0wvaagLKAVWq8bbeNwnZZK1r1XQysX3xurLU">
  <meta name="google-site-verification" content="ZzhVyEFwb7w3e0-uOTltm8Jsck2F5StVihD0exw2fsA">
  <meta name="google-site-verification" content="GXs5KoUUkNCoaAZn7wPN-t01Pywp9M3sEjnt_3_ZWPc">

<meta name="octolytics-host" content="collector.githubapp.com" /><meta name="octolytics-app-id" content="github" /><meta name="octolytics-event-url" content="https://collector.githubapp.com/github-external/browser_event" /><meta name="octolytics-dimension-ga_id" content="" class="js-octo-ga-id" /><meta name="octolytics-actor-id" content="62664434" /><meta name="octolytics-actor-login" content="monninn82" /><meta name="octolytics-actor-hash" content="bd18f112bd271b26e147b2e1fabdedef5b9da9c0c6270a6b4eeeb4042803d3ad" />
<meta name="analytics-location" content="/&lt;user-name&gt;/&lt;repo-name&gt;/blob/show" data-pjax-transient="true" />



    <meta name="google-analytics" content="UA-3769691-2">

  <meta class="js-ga-set" name="userId" content="34f66778b90a2c84fd58e7d29e9833e3">

<meta class="js-ga-set" name="dimension1" content="Logged In">



  

      <meta name="hostname" content="github.com">
    <meta name="user-login" content="monninn82">

      <meta name="expected-hostname" content="github.com">

      <meta name="js-proxy-site-detection-payload" content="ZjQ3NmI3NmQzZDFiNTFkNTg2ZTVmMTdmOGVmM2I4ZjI2YTE2NGM1OGM4YTBkNGQ0ODM2M2MzMzVkYmM5OTdmOXx7InJlbW90ZV9hZGRyZXNzIjoiMTA0LjEyOS4xOTYuNzYiLCJyZXF1ZXN0X2lkIjoiMzIzRDo2RjJDOjUzMzlBMjo4NjI0M0E6NUU4OTZFQUEiLCJ0aW1lc3RhbXAiOjE1ODYwNjUwNzksImhvc3QiOiJnaXRodWIuY29tIn0=">

    <meta name="enabled-features" content="MARKETPLACE_FEATURED_BLOG_POSTS,MARKETPLACE_INVOICED_BILLING,MARKETPLACE_SOCIAL_PROOF_CUSTOMERS,MARKETPLACE_TRENDING_SOCIAL_PROOF,MARKETPLACE_RECOMMENDATIONS,MARKETPLACE_PENDING_INSTALLATIONS,RELATED_ISSUES,GHE_CLOUD_TRIAL,PAGE_STALE_CHECK">

  <meta http-equiv="x-pjax-version" content="80ebc3286fe9d88acaddc1320106dfc2">
  

      <link href="https://github.com/matiassingers/awesome-readme/commits/master.atom" rel="alternate" title="Recent Commits to awesome-readme:master" type="application/atom+xml">

  <meta name="go-import" content="github.com/matiassingers/awesome-readme git https://github.com/matiassingers/awesome-readme.git">

  <meta name="octolytics-dimension-user_id" content="938453" /><meta name="octolytics-dimension-user_login" content="matiassingers" /><meta name="octolytics-dimension-repository_id" content="29456624" /><meta name="octolytics-dimension-repository_nwo" content="matiassingers/awesome-readme" /><meta name="octolytics-dimension-repository_public" content="true" /><meta name="octolytics-dimension-repository_is_fork" content="false" /><meta name="octolytics-dimension-repository_network_root_id" content="29456624" /><meta name="octolytics-dimension-repository_network_root_nwo" content="matiassingers/awesome-readme" /><meta name="octolytics-dimension-repository_explore_github_marketplace_ci_cta_shown" content="false" />


    <link rel="canonical" href="https://github.com/matiassingers/awesome-readme/blob/master/readme.md" data-pjax-transient>


  <meta name="browser-stats-url" content="https://api.github.com/_private/browser/stats">

  <meta name="browser-errors-url" content="https://api.github.com/_private/browser/errors">

  <link rel="mask-icon" href="https://github.githubassets.com/pinned-octocat.svg" color="#000000">
  <link rel="alternate icon" class="js-site-favicon" type="image/png" href="https://github.githubassets.com/favicons/favicon.png">
  <link rel="icon" class="js-site-favicon" type="image/svg+xml" href="https://github.githubassets.com/favicons/favicon.svg">

<meta name="theme-color" content="#1e2327">


  <link rel="manifest" href="/manifest.json" crossOrigin="use-credentials">

  </head>

  <body class="logged-in env-production page-responsive page-blob">
    

  <div class="position-relative js-header-wrapper ">
    <a href="#start-of-content" class="p-3 bg-blue text-white show-on-focus js-skip-to-content">Skip to content</a>
    <span class="Progress progress-pjax-loader position-fixed width-full js-pjax-loader-bar">
      <span class="progress-pjax-loader-bar top-0 left-0" style="width: 0%;"></span>
    </span>

    
    



        <header class="Header js-details-container Details flex-wrap flex-lg-nowrap p-responsive" role="banner">

  <div class="Header-item d-none d-lg-flex">
    <a class="Header-link" href="https://github.com/" data-hotkey="g d"
  aria-label="Homepage " data-ga-click="Header, go to dashboard, icon:logo">
  <svg class="octicon octicon-mark-github v-align-middle" height="32" viewBox="0 0 16 16" version="1.1" width="32" aria-hidden="true"><path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0016 8c0-4.42-3.58-8-8-8z"></path></svg>
</a>

  </div>

  <div class="Header-item d-lg-none">
    <button class="Header-link btn-link js-details-target" type="button" aria-label="Toggle navigation" aria-expanded="false">
      <svg height="24" class="octicon octicon-three-bars" viewBox="0 0 12 16" version="1.1" width="18" aria-hidden="true"><path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path></svg>
    </button>
  </div>

  <div class="Header-item Header-item--full flex-column flex-lg-row width-full flex-order-2 flex-lg-order-none mr-0 mr-lg-3 mt-3 mt-lg-0 Details-content--hidden">
      <div class="header-search flex-self-stretch flex-lg-self-auto mr-0 mr-lg-3 mb-3 mb-lg-0 scoped-search site-scoped-search js-site-search position-relative js-jump-to"
  role="combobox"
  aria-owns="jump-to-results"
  aria-label="Search or jump to"
  aria-haspopup="listbox"
  aria-expanded="false"
>
  <div class="position-relative">
    <!-- '"` --><!-- </textarea></xmp> --></option></form><form class="js-site-search-form" role="search" aria-label="Site" data-scope-type="Repository" data-scope-id="29456624" data-scoped-search-url="/matiassingers/awesome-readme/search" data-unscoped-search-url="/search" action="/matiassingers/awesome-readme/search" accept-charset="UTF-8" method="get">
      <label class="form-control input-sm header-search-wrapper p-0 header-search-wrapper-jump-to position-relative d-flex flex-justify-between flex-items-center js-chromeless-input-container">
        <input type="text"
          class="form-control input-sm header-search-input jump-to-field js-jump-to-field js-site-search-focus js-site-search-field is-clearable"
          data-hotkey="s,/"
          name="q"
          value=""
          placeholder="Search or jump to…"
          data-unscoped-placeholder="Search or jump to…"
          data-scoped-placeholder="Search or jump to…"
          autocapitalize="off"
          aria-autocomplete="list"
          aria-controls="jump-to-results"
          aria-label="Search or jump to…"
          data-jump-to-suggestions-path="/_graphql/GetSuggestedNavigationDestinations"
          spellcheck="false"
          autocomplete="off"
          >
          <input type="hidden" value="2Zp+I4m+Ol3N5Uva2wfUJ/v+wr/ctHwRjJMO++/jqHVEe6D8CGqUan3xDleNMFH40t46eDuXtyPgxPKOskjbgw==" data-csrf="true" class="js-data-jump-to-suggestions-path-csrf" />
          <input type="hidden" class="js-site-search-type-field" name="type" >
            <img src="https://github.githubassets.com/images/search-key-slash.svg" alt="" class="mr-2 header-search-key-slash">

            <div class="Box position-absolute overflow-hidden d-none jump-to-suggestions js-jump-to-suggestions-container">
              
<ul class="d-none js-jump-to-suggestions-template-container">
  

<li class="d-flex flex-justify-start flex-items-center p-0 f5 navigation-item js-navigation-item js-jump-to-suggestion" role="option">
  <a tabindex="-1" class="no-underline d-flex flex-auto flex-items-center jump-to-suggestions-path js-jump-to-suggestion-path js-navigation-open p-2" href="">
    <div class="jump-to-octicon js-jump-to-octicon flex-shrink-0 mr-2 text-center d-none">
      <svg height="16" width="16" class="octicon octicon-repo flex-shrink-0 js-jump-to-octicon-repo d-none" title="Repository" aria-label="Repository" viewBox="0 0 12 16" version="1.1" role="img"><path fill-rule="evenodd" d="M4 9H3V8h1v1zm0-3H3v1h1V6zm0-2H3v1h1V4zm0-2H3v1h1V2zm8-1v12c0 .55-.45 1-1 1H6v2l-1.5-1.5L3 16v-2H1c-.55 0-1-.45-1-1V1c0-.55.45-1 1-1h10c.55 0 1 .45 1 1zm-1 10H1v2h2v-1h3v1h5v-2zm0-10H2v9h9V1z"></path></svg>
      <svg height="16" width="16" class="octicon octicon-project flex-shrink-0 js-jump-to-octicon-project d-none" title="Project" aria-label="Project" viewBox="0 0 15 16" version="1.1" role="img"><path fill-rule="evenodd" d="M10 12h3V2h-3v10zm-4-2h3V2H6v8zm-4 4h3V2H2v12zm-1 1h13V1H1v14zM14 0H1a1 1 0 00-1 1v14a1 1 0 001 1h13a1 1 0 001-1V1a1 1 0 00-1-1z"></path></svg>
      <svg height="16" width="16" class="octicon octicon-search flex-shrink-0 js-jump-to-octicon-search d-none" title="Search" aria-label="Search" viewBox="0 0 16 16" version="1.1" role="img"><path fill-rule="evenodd" d="M15.7 13.3l-3.81-3.83A5.93 5.93 0 0013 6c0-3.31-2.69-6-6-6S1 2.69 1 6s2.69 6 6 6c1.3 0 2.48-.41 3.47-1.11l3.83 3.81c.19.2.45.3.7.3.25 0 .52-.09.7-.3a.996.996 0 000-1.41v.01zM7 10.7c-2.59 0-4.7-2.11-4.7-4.7 0-2.59 2.11-4.7 4.7-4.7 2.59 0 4.7 2.11 4.7 4.7 0 2.59-2.11 4.7-4.7 4.7z"></path></svg>
    </div>

    <img class="avatar mr-2 flex-shrink-0 js-jump-to-suggestion-avatar d-none" alt="" aria-label="Team" src="" width="28" height="28">

    <div class="jump-to-suggestion-name js-jump-to-suggestion-name flex-auto overflow-hidden text-left no-wrap css-truncate css-truncate-target">
    </div>

    <div class="border rounded-1 flex-shrink-0 bg-gray px-1 text-gray-light ml-1 f6 d-none js-jump-to-badge-search">
      <span class="js-jump-to-badge-search-text-default d-none" aria-label="in this repository">
        In this repository
      </span>
      <span class="js-jump-to-badge-search-text-global d-none" aria-label="in all of GitHub">
        All GitHub
      </span>
      <span aria-hidden="true" class="d-inline-block ml-1 v-align-middle">↵</span>
    </div>

    <div aria-hidden="true" class="border rounded-1 flex-shrink-0 bg-gray px-1 text-gray-light ml-1 f6 d-none d-on-nav-focus js-jump-to-badge-jump">
      Jump to
      <span class="d-inline-block ml-1 v-align-middle">↵</span>
    </div>
  </a>
</li>

</ul>

<ul class="d-none js-jump-to-no-results-template-container">
  <li class="d-flex flex-justify-center flex-items-center f5 d-none js-jump-to-suggestion p-2">
    <span class="text-gray">No suggested jump to results</span>
  </li>
</ul>

<ul id="jump-to-results" role="listbox" class="p-0 m-0 js-navigation-container jump-to-suggestions-results-container js-jump-to-suggestions-results-container">
  

<li class="d-flex flex-justify-start flex-items-center p-0 f5 navigation-item js-navigation-item js-jump-to-scoped-search d-none" role="option">
  <a tabindex="-1" class="no-underline d-flex flex-auto flex-items-center jump-to-suggestions-path js-jump-to-suggestion-path js-navigation-open p-2" href="">
    <div class="jump-to-octicon js-jump-to-octicon flex-shrink-0 mr-2 text-center d-none">
      <svg height="16" width="16" class="octicon octicon-repo flex-shrink-0 js-jump-to-octicon-repo d-none" title="Repository" aria-label="Repository" viewBox="0 0 12 16" version="1.1" role="img"><path fill-rule="evenodd" d="M4 9H3V8h1v1zm0-3H3v1h1V6zm0-2H3v1h1V4zm0-2H3v1h1V2zm8-1v12c0 .55-.45 1-1 1H6v2l-1.5-1.5L3 16v-2H1c-.55 0-1-.45-1-1V1c0-.55.45-1 1-1h10c.55 0 1 .45 1 1zm-1 10H1v2h2v-1h3v1h5v-2zm0-10H2v9h9V1z"></path></svg>
      <svg height="16" width="16" class="octicon octicon-project flex-shrink-0 js-jump-to-octicon-project d-none" title="Project" aria-label="Project" viewBox="0 0 15 16" version="1.1" role="img"><path fill-rule="evenodd" d="M10 12h3V2h-3v10zm-4-2h3V2H6v8zm-4 4h3V2H2v12zm-1 1h13V1H1v14zM14 0H1a1 1 0 00-1 1v14a1 1 0 001 1h13a1 1 0 001-1V1a1 1 0 00-1-1z"></path></svg>
      <svg height="16" width="16" class="octicon octicon-search flex-shrink-0 js-jump-to-octicon-search d-none" title="Search" aria-label="Search" viewBox="0 0 16 16" version="1.1" role="img"><path fill-rule="evenodd" d="M15.7 13.3l-3.81-3.83A5.93 5.93 0 0013 6c0-3.31-2.69-6-6-6S1 2.69 1 6s2.69 6 6 6c1.3 0 2.48-.41 3.47-1.11l3.83 3.81c.19.2.45.3.7.3.25 0 .52-.09.7-.3a.996.996 0 000-1.41v.01zM7 10.7c-2.59 0-4.7-2.11-4.7-4.7 0-2.59 2.11-4.7 4.7-4.7 2.59 0 4.7 2.11 4.7 4.7 0 2.59-2.11 4.7-4.7 4.7z"></path></svg>
    </div>

    <img class="avatar mr-2 flex-shrink-0 js-jump-to-suggestion-avatar d-none" alt="" aria-label="Team" src="" width="28" height="28">

    <div class="jump-to-suggestion-name js-jump-to-suggestion-name flex-auto overflow-hidden text-left no-wrap css-truncate css-truncate-target">
    </div>


docs/CHANGELOG.md
# [1.18.0](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.17.0...v1.18.0) (2020-03-12)


### Features

* Return executeData in response to EXECUTE lifecycle event ([#141](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/141)) ([6aa0340](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/6aa0340))

# [1.17.0](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.16.0...v1.17.0) (2020-03-04)


### Features

* Removed auto-confirmation for API Access apps ([#142](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/142)) ([43694a2](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/43694a2))

## [1.15.1](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.15.0...v1.15.1) (2020-01-02)


### Bug Fixes

* package.json, package-lock.json & .snyk to reduce vulnerabilities ([#130](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/130)) ([600cede](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/600cede))

# [1.15.0](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.14.2...v1.15.0) (2019-12-24)


### Features

* Added support for rules API ([#133](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/133)) ([738a6da](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/738a6da))

## [1.14.2](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.14.1...v1.14.2) (2019-12-24)


### Bug Fixes

* reverted text-setting change at [#127](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/127) as [#131](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/131) merged ([#134](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/134)) ([ab57673](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/ab57673))

## [1.14.1](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.14.0...v1.14.1) (2019-12-23)


### Bug Fixes

* applied translateDefaultValue to SectionSettings to ensure compatibility ([#131](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/131)) ([02b3bfb](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/02b3bfb))

# [1.14.0](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.13.6...v1.14.0) (2019-12-10)


### Features

* Added switch to disable translating of options ([#129](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/129)) ([82c0d54](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/82c0d54))

## [1.13.6](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.13.5...v1.13.6) (2019-12-09)


### Bug Fixes

* Log responses as well as events for lambdas as is done for webhooks ([#128](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/128)) ([929f401](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/929f401))

## [1.13.5](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.13.4...v1.13.5) (2019-10-14)


### Bug Fixes

* Corrected bug in automatic refresh of tokens ([#120](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/120)) ([affa856](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/affa856))

## [1.13.4](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.13.3...v1.13.4) (2019-10-08)


### Bug Fixes

* update dependencies ([#119](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/119)) ([2e29498](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/2e29498))

## [1.13.3](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.13.2...v1.13.3) (2019-10-08)


### Bug Fixes

* package.json & .snyk to reduce vulnerabilities ([#118](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/118)) ([841cec9](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/841cec9))

## [1.13.2](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.13.1...v1.13.2) (2019-09-26)


### Bug Fixes

* release semaphore when sequential api throws non-401 errors ([#117](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/117)) ([2fa347e](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/2fa347e))

## [1.13.1](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.13.0...v1.13.1) (2019-09-24)


### Bug Fixes

* catch 404 exception from device health for unenrolled devices ([#116](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/116)) ([f8f3ba5](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/f8f3ba5))

# [1.13.0](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.12.0...v1.13.0) (2019-09-24)


### Features

* Pass event time to handlers ([#115](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/115)) ([d71e259](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/d71e259))

# [1.12.0](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.11.0...v1.12.0) (2019-09-24)


### Features

* EnumSettings i18n ([#110](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/110)) ([fb82636](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/fb82636))

# [1.11.0](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.10.1...v1.11.0) (2019-09-17)


### Features

* Change to be able to set localized postMessage in decimal-setting ([#114](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/issues/114)) ([824674a](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/commit/824674a))

## [1.10.1](https://github.com/SmartThingsCommunity/smartapp-sdk-nodejs/compare/v1.10.0...v1.10.1) (2019-09-06)


### Bug Fixes


alarmserver/devices.py
#!/usr/bin/python2.7

import asyncore, asynchat
import ConfigParser
import datetime
import os, socket, string, sys, httplib, urllib, urlparse, ssl
import StringIO, mimetools
import json
import hashlib
import time
import getopt
import requests


MAXZONES=128
MAXPARTITIONS=16

class AlarmServerConfig():
    def __init__(self, configfile):

        self._config = ConfigParser.ConfigParser()
        self._config.read(configfile)

        self.LOGURLREQUESTS = self.read_config_var('alarmserver', 'logurlrequests', True, 'bool')
        self.HTTPSPORT = self.read_config_var('alarmserver', 'httpsport', 8111, 'int')
        self.CERTFILE = self.read_config_var('alarmserver', 'certfile', 'server.crt', 'str')
        self.KEYFILE = self.read_config_var('alarmserver', 'keyfile', 'server.key', 'str')
        self.MAXEVENTS = self.read_config_var('alarmserver', 'maxevents', 10, 'int')
        self.MAXALLEVENTS = self.read_config_var('alarmserver', 'maxallevents', 100, 'int')
        self.ENVISALINKHOST = self.read_config_var('envisalink', 'host', 'envisalink', 'str')
        self.ENVISALINKPORT = self.read_config_var('envisalink', 'port', 4025, 'int')
        self.ENVISALINKPASS = self.read_config_var('envisalink', 'pass', 'user', 'str')
        self.ENABLEPROXY = self.read_config_var('envisalink', 'enableproxy', True, 'bool')
        self.ENVISALINKPROXYPORT = self.read_config_var('envisalink', 'proxyport', self.ENVISALINKPORT, 'int')
        self.ENVISALINKPROXYPASS = self.read_config_var('envisalink', 'proxypass', self.ENVISALINKPASS, 'str')
        self.PUSHOVER_ENABLE = self.read_config_var('pushover', 'enable', False, 'bool')
        self.PUSHOVER_USERTOKEN = self.read_config_var('pushover', 'enable', False, 'bool')
        self.ALARMCODE = self.read_config_var('envisalink', 'alarmcode', 1111, 'str')
        self.EVENTTIMEAGO = self.read_config_var('alarmserver', 'eventtimeago', True, 'bool')
        self.LOGFILE = self.read_config_var('alarmserver', 'logfile', '', 'str')
        self.CALLBACKURL_BASE = self.read_config_var('alarmserver', 'callbackurl_base', '', 'str')
        self.CALLBACKURL_APP_ID = self.read_config_var('alarmserver', 'callbackurl_app_id', '', 'str')
        self.CALLBACKURL_ACCESS_TOKEN = self.read_config_var('alarmserver', 'callbackurl_access_token', '', 'str')
        self.CALLBACKURL_EVENT_CODES = self.read_config_var('alarmserver', 'callbackurl_event_codes', '', 'str')
        global LOGTOFILE
        if self.LOGFILE == '':
            LOGTOFILE = False
        else:
            LOGTOFILE = True

        self.PARTITIONNAMES={}
        self.PARTITIONS={}
        for i in range(1, MAXPARTITIONS+1):
            self.PARTITIONNAMES[i]=self.read_config_var('partition'+str(i), 'name', False, 'str', True)
            stay=self.read_config_var('partition'+str(i), 'stay', False, 'str', True)
            away=self.read_config_var('partition'+str(i), 'away', False, 'str', True)
            if stay!=False or away!=False:
                self.PARTITIONS[i] = {}
                if away!=False:
                    self.PARTITIONS[i]['away']=away
                if stay!=False:
                    self.PARTITIONS[i]['stay']=stay

        self.ZONES={}
        self.ZONENAMES={}
        for i in range(1, MAXZONES+1):
            self.ZONENAMES[i]=self.read_config_var('zone'+str(i), 'name', False, 'str', True)
            type = self.read_config_var('zone'+str(i), 'type', False, 'str', True)
            if(self.ZONENAMES[i]!=False and type!=False):
                self.ZONES[i] = {}
                self.ZONES[i]['name'] = self.ZONENAMES[i]
                self.ZONES[i]['type'] = type

    def defaulting(self, section, variable, default, quiet = False):
        if quiet == False:
            print('Config option '+ str(variable) + ' not set in ['+str(section)+'] defaulting to: \''+str(default)+'\'')

    def read_config_var(self, section, variable, default, type = 'str', quiet = False):
        try:
            if type == 'str':
                return self._config.get(section,variable)
            elif type == 'bool':
                return self._config.getboolean(section,variable)
            elif type == 'int':
                return int(self._config.get(section,variable))
        except (ConfigParser.NoSectionError, ConfigParser.NoOptionError):
            self.defaulting(section, variable, default, quiet)
            return default

class DeviceSetup():
    def __init__(self, config, delete):
        self._config = config

        # prepare partition and zone json for device creation
        if delete:
            partitionjson = json.dumps({})
            zonejson = json.dumps({})
        else:
            partitionjson = json.dumps(config.PARTITIONS)
            zonejson = json.dumps(config.ZONES)

        headers = {'content-type': 'application/json'}

        # create zone devices
        myURL = config.CALLBACKURL_BASE + '/' + config.CALLBACKURL_APP_ID + '/installzones' + '?access_token=' + config.CALLBACKURL_ACCESS_TOKEN
        print('myURL: %s' % myURL)
        r = requests.post(myURL, data=zonejson, headers=headers)
        print('zonejson: %s\nStatus: %s\nMessage: %s' % (zonejson, r.status_code, r.text))

        # create partition devices
        myURL = config.CALLBACKURL_BASE + '/' + config.CALLBACKURL_APP_ID + '/installpartitions' + '?access_token=' + config.CALLBACKURL_ACCESS_TOKEN
        print('myURL: %s' % myURL)
        r = requests.post(myURL, data=partitionjson, headers=headers)
        print('partitionjson: %s\nStatus: %s\nMessage: %s' % (partitionjson, r.status_code, r.text))



def usage():
    print 'Usage: '+sys.argv[0]+' -c <configfile>'


def main(argv):
    try:
      opts, args = getopt.getopt(argv, "hdc:", ["help", "config="])
    except getopt.GetoptError:
        usage()
        sys.exit(2)
    for opt, arg in opts:
        if opt in ("-h", "--help"):
            usage()
            sys.exit()
        elif opt in ("-c", "--config"):
            global conffile
            conffile = arg
        elif opt in ("-d", "--delete"):
            global delete
            delete = True


if __name__=="__main__":
    conffile='alarmserver.cfg'
    delete=False
    main(sys.argv[1:])
    print('Using configuration file %s' % conffile)
    config = AlarmServerConfig(conffile)
    DeviceSetup(config, delete)
